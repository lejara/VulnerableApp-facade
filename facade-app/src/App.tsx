import React from "react";
import "./App.css";
import {
  Container as RSuiteContainer,
  Sidebar as RSuitesSidebar,
} from "rsuite";
import "rsuite/dist/styles/rsuite-default.min.css";
import Header from "./Components/Header";
import LeftNav from "./Components/LeftNav";
import { GlobalState } from "./interface/State";
import Footer from "./Components/Footer";
import { Content } from "./Components/Content";
import { getResource } from "./Utilities/Utils";
import { VulnerabilityDefinitionResponse } from "./interface/GeneralContracts";

export default class App extends React.Component {
  state: GlobalState = {
    isSuccessfullyLoaded: false,
    activateAboutUsPage: false,
    activateHomePage: true,
    showHints: false,
  };

  _getResourcesInformationsForLevel(levelInformationRes: any) {
    let resourceInformationHtml = {
      isAbsolute:
        levelInformationRes["resourceInformation"]["htmlResource"][
          "isAbsolute"
        ],
      uri: levelInformationRes["resourceInformation"]["htmlResource"]["uri"],
    };

    let staticResourceInformation = [];
    for (let staticResourceInformationIndexRes in levelInformationRes[
      "levels"
    ]) {
      let staticResourceInformationRes =
        levelInformationRes["levels"][staticResourceInformationIndexRes];
      staticResourceInformation.push({
        isAbsolute: staticResourceInformationRes["isAbsolute"],
        uri: staticResourceInformationRes["uri"],
        resourceType: staticResourceInformationRes["resourceType"],
      });
    }
    return {
      resourceInformationHtml: resourceInformationHtml,
      staticResourceInformation: staticResourceInformation,
    };
  }

  _getHintsForLevel(levelInformationRes: any) {
    let hints = [];
    for (let hintIndexRes in levelInformationRes["hints"]) {
      let hintRes = levelInformationRes["hints"][hintIndexRes];
      let vulnerabilityTypes = [];
      for (let vulnerabilityTypeIndexRes in hintRes["vulnerabilityTypes"]) {
        vulnerabilityTypes.push({
          identifierType:
            hintRes["vulnerabilityTypes"][vulnerabilityTypeIndexRes][
              "identifierType"
            ],
          value:
            hintRes["vulnerabilityTypes"][vulnerabilityTypeIndexRes]["value"],
        });
      }
      let hint = {
        description: hintRes[hintIndexRes],
        vulnerabilityTypes: vulnerabilityTypes,
      };
      hints.push(hint);
    }
    return hints;
  }

  _getVulnerabilityTypesForLevel(vulnerabilityDefinitionsRes: any) {
    let vulnerabilityTypes = [];
    for (let vulnerabilityTypeIndexRes in vulnerabilityDefinitionsRes[
      "vulnerabilityTypes"
    ]) {
      vulnerabilityTypes.push({
        identifierType:
          vulnerabilityDefinitionsRes["vulnerabilityTypes"][
            vulnerabilityTypeIndexRes
          ]["identifierType"],
        value:
          vulnerabilityDefinitionsRes["vulnerabilityTypes"][
            vulnerabilityTypeIndexRes
          ]["value"],
      });
    }
    return vulnerabilityTypes;
  }

  _populateLevelsForVulnerability(
    vulnerabilityDefinitionsRes: any,
    vulnerabilityDefinitionIndexRes: string
  ) {
    let levels = [];
    for (let vulnerabilityLevelIndexRes in vulnerabilityDefinitionsRes[
      vulnerabilityDefinitionIndexRes
    ]["levels"]) {
      let levelInformationRes =
        vulnerabilityDefinitionsRes[vulnerabilityDefinitionIndexRes]["levels"][
          vulnerabilityLevelIndexRes
        ];
      //Hint population
      let hints = this._getHintsForLevel(levelInformationRes);
      // Resource population
      let resourceInformations =
        this._getResourcesInformationsForLevel(levelInformationRes);

      let levelInformation = {
        levelIdentifier: levelInformationRes["levelIdentifier"],
        variant: levelInformationRes["variant"],
        hints: hints,
        resourceInformation: {
          htmlResource: resourceInformations.resourceInformationHtml,
          staticResources: resourceInformations.staticResourceInformation,
        },
      };
      levels.push(levelInformation);
      return levels;
    }
  }

  _populateGlobalState(
    vulnerabilityDefinitionResponse: VulnerabilityDefinitionResponse
  ) {
    if (vulnerabilityDefinitionResponse.isSuccessful) {
      if (!vulnerabilityDefinitionResponse.data) {
        return;
      }
      let applicationsDataRes = vulnerabilityDefinitionResponse.data;
      let applicationDataArray = [];
      for (let vulnerableAppRes in applicationsDataRes) {
        let vulnerabilityDefinitionsRes = applicationsDataRes[vulnerableAppRes];
        let vulnerabilityDefinitions = [];
        for (let vulnerabilityDefinitionIndexRes in vulnerabilityDefinitionsRes) {
          let levels = this._populateLevelsForVulnerability(
            vulnerabilityDefinitionsRes,
            vulnerabilityDefinitionIndexRes
          );
          let vulnerabilityTypes = this._getVulnerabilityTypesForLevel(
            vulnerabilityDefinitionIndexRes
          );
          let vulnerabilityDefinition = {
            name: vulnerabilityDefinitionsRes["name"],
            id: vulnerabilityDefinitionsRes["id"],
            description: vulnerabilityDefinitionsRes["description"],
            vulnerabilityTypes: vulnerabilityTypes,
            levels: levels,
          };
          vulnerabilityDefinitions.push(vulnerabilityDefinition);
        }
        applicationDataArray.push({
          applicationName: vulnerableAppRes,
          vulnerabilityDefinitions: vulnerabilityDefinitionsRes,
        });
      }
      this.setState({
        applicationData: applicationDataArray,
        activeApplication:
          applicationDataArray.length >= 1
            ? applicationDataArray[0].applicationName
            : null,
        isSuccessfullyLoaded: true,
      });
    } else {
      console.log(vulnerabilityDefinitionResponse.error);
    }
  }

  componentDidMount() {
    getResource(
      "/VulnerabilityDefinitions",
      this._populateGlobalState.bind(this),
      true
    );
  }

  setGlobalState = (globalState: GlobalState) => {
    this.setState(globalState);
    //Sets the global state such that the vulnerable applications
    //knows the execution context details which can help them write the
    //templates.
    (window as any).globalUtilityState = {
      activeVulnerabilityIdentifier: globalState.activeVulnerability,
      activeVulnerabilityLevelIdentifier: globalState.activeLevel,
      baseURI: window.location.toString(),
    };
  };

  render() {
    const copyRightYear = new Date().getFullYear();

    return (
      <RSuiteContainer className="show-container">
        <Header
          globalState={this.state}
          setGlobalState={this.setGlobalState}
        ></Header>
        <RSuiteContainer
          className="show-container"
          style={{ display: "flex", width: "100vw" }}
        >
          <RSuitesSidebar>
            <LeftNav
              globalState={this.state}
              setGlobalState={this.setGlobalState}
            />
          </RSuitesSidebar>
          <Content
            globalState={this.state}
            setGlobalState={this.setGlobalState}
          ></Content>
        </RSuiteContainer>
        <Footer copyRightYear={copyRightYear}>Footer</Footer>
      </RSuiteContainer>
    );
  }
}
