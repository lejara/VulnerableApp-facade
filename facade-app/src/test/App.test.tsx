import React from "react";
import { render } from "@testing-library/react";

import App from "../App";
import Header from "../Components/Header";

describe("App", () => {
  it("renders correctly", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
  // it("Calls at least one vul", () => {
  //   const spy = jest.spyOn(App.prototype, "_populateLevelsForVulnerability");
  //   render(<App />);
  //   expect(spy).toHaveBeenCalledTimes(1);
  // });
  // it("renders correctly", () => {
  //   const mock = () => jest.fn();
  //   const state = {
  //     isSuccessfullyLoaded: false,
  //     activateAboutUsPage: false,
  //     activateHomePage: true,
  //     showHints: false,
  //   };
  //   const { container } = render(
  //     <Header setGlobalState={mock} globalState={state} />
  //   );
  //   expect(container).toMatchSnapshot();
  // });
});
