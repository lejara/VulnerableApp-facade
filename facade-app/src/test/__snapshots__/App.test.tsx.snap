// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <div
    class="rs-container show-container"
  >
    <div
      class="rs-header"
    >
      <div
        class="rs-navbar rs-navbar-inverse"
        role="navigation"
      >
        <div
          class="rs-navbar-header"
        >
          <img
            alt="vulnerable app logo"
            height="55"
            src="Logo.png"
            width="55"
          />
        </div>
        <div
          class="rs-navbar-body"
        >
          <div
            class="rs-nav rs-nav-default rs-navbar-nav rs-nav-horizontal"
          >
            <ul>
              <li
                class="rs-nav-item"
              >
                <a
                  class="rs-nav-item-content"
                  role="button"
                  tabindex="0"
                >
                  <b>
                    Owasp VulnerableApp-Facade
                  </b>
                  <span
                    class="rs-ripple-pond"
                  >
                    <span
                      class="rs-ripple"
                    />
                  </span>
                </a>
              </li>
            </ul>
          </div>
          <div
            class="rs-nav rs-nav-default rs-navbar-nav rs-navbar-right rs-nav-horizontal"
          >
            <ul>
              <li
                class="rs-nav-item"
              >
                <a
                  class="rs-nav-item-content"
                  role="button"
                  tabindex="0"
                >
                  <i
                    class="rs-icon rs-icon-home"
                    role="img"
                  />
                  Home
                  <span
                    class="rs-ripple-pond"
                  >
                    <span
                      class="rs-ripple"
                    />
                  </span>
                </a>
              </li>
              <li
                class="rs-nav-item"
              >
                <a
                  class="rs-nav-item-content"
                  role="button"
                  tabindex="0"
                >
                  About Us
                  <span
                    class="rs-ripple-pond"
                  >
                    <span
                      class="rs-ripple"
                    />
                  </span>
                </a>
              </li>
              <a
                href="https://github.com/SasanLabs/VulnerableApp-facade"
              >
                <li
                  class="rs-nav-item"
                >
                  <a
                    class="rs-nav-item-content"
                    role="button"
                    tabindex="0"
                  >
                    <i
                      class="rs-icon rs-icon-github"
                      role="img"
                    />
                    Github
                    <span
                      class="rs-ripple-pond"
                    >
                      <span
                        class="rs-ripple"
                      />
                    </span>
                  </a>
                </li>
              </a>
              <li
                class="rs-dropdown rs-dropdown-placement-bottom-start"
                role="menu"
              >
                <ul
                  class="rs-dropdown-menu"
                  role="menu"
                >
                  <a
                    href="https://github.com/SasanLabs/VulnerableApp"
                  >
                    <li
                      class="rs-dropdown-item rs-dropdown-item-with-icon"
                    >
                      <a
                        class="rs-dropdown-item-content"
                        tabindex="-1"
                      >
                        <i
                          class="rs-dropdown-item-menu-icon rs-icon rs-icon-github"
                          role="img"
                        />
                        Owasp VulnerableApp
                      </a>
                    </li>
                  </a>
                  <a
                    href="https://github.com/SasanLabs/owasp-zap-jwt-addon"
                  >
                     
                    <li
                      class="rs-dropdown-item rs-dropdown-item-with-icon"
                    >
                      <a
                        class="rs-dropdown-item-content"
                        tabindex="-1"
                      >
                        <i
                          class="rs-dropdown-item-menu-icon rs-icon rs-icon-github"
                          role="img"
                        />
                        ZAP JWT Addon
                      </a>
                    </li>
                  </a>
                  <a
                    href="https://github.com/SasanLabs/owasp-zap-fileupload-addon"
                  >
                    <li
                      class="rs-dropdown-item rs-dropdown-item-with-icon"
                    >
                      <a
                        class="rs-dropdown-item-content"
                        tabindex="-1"
                      >
                        <i
                          class="rs-dropdown-item-menu-icon rs-icon rs-icon-github"
                          role="img"
                        />
                        ZAP FileUpload Addon
                      </a>
                    </li>
                  </a>
                </ul>
                <a
                  class="rs-btn rs-btn-subtle rs-dropdown-toggle"
                  tabindex="0"
                >
                  Projects by SasanLabs
                  <span
                    class="rs-dropdown-toggle-caret"
                  />
                  <span
                    class="rs-ripple-pond"
                  >
                    <span
                      class="rs-ripple"
                    />
                  </span>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      class="rs-container show-container rs-container-has-sidebar"
      style="display: flex; width: 100vw;"
    >
      <div
        class="rs-sidebar"
        style="flex: 0 0 260px; width: 260px;"
      >
        <div />
      </div>
      <div
        class="VulnerableApp-Facade-Info"
      >
        <div>
          <div
            class="rs-panel-group"
          >
            <div
              data-testid="HOME_PAGE_MAIN_CONTAINER"
            >
              <div
                class="VulnerableApp-Facade-HomePage-Content-Header rs-panel rs-panel-default"
              >
                <div
                  class="rs-panel-heading"
                  role="rowheader"
                  tabindex="-1"
                >
                  Welcome to Owasp VulnerableApp-Facade
                </div>
                <div
                  class="rs-panel-body"
                >
                  <div
                    class="VulnerableApp-Facade-HomePage-Content"
                    data-testid="HOME_PAGE_CONTENT_CONTAINER"
                  >
                    As we are seeing a lot of technological enhancements in the industry from past few years, these technical enhancements are solving one or the other problem however, with that they also bring few different vulnerabilities. Vulnerable Applications are generally written in one of the techstacks like either Node.js or Java with a SQL or NoSQL database etc and hence they are not able to expand to a whole new set of vulnerabilities which are present in other technologies. Also adding more vulnerabilities in a single vulnerable application makes it heavier and complex which finally makes it unmaintainable. So VulnerableApp-facade is built to solve this problem by building a distributed farm of Vulnerable Applications such that they can be built agnostic to tech stacks.
                    <p>
                      Following is the design diagram of Owasp VulnerableApp-Facade:
                    </p>
                    <img
                      alt="Unable to lead the design diagram"
                      src="/images/VulnerableApp-facade.jpeg"
                      width="80%"
                    />
                    <p>
                      Here VulnerableApp-Facade is running as a gateway or a proxy which is routing calls to actual Vulnerable Applications based on a criteria defined in the nginx configuration.
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="VulnerableApp-Facade-HomePage-Content-Header rs-panel rs-panel-default"
                data-testid="HOME_PAGE_WARNING_CONTAINER"
              >
                <div
                  class="rs-panel-heading"
                  role="rowheader"
                  tabindex="-1"
                >
                  Warning
                </div>
                <div
                  class="rs-panel-body"
                >
                  <div
                    class="VulnerableApp-Facade-HomePage-Content"
                  >
                    As VulnerableApp-Facade is a proxy wrapper over the actual Vulnerable Applications which can be very dangerous if exposed over the public internet. So we suggest you to please run it in local dev environments or environments without any public internet access.
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="rs-footer"
    >
      <div
        class="rs-navbar rs-navbar-default"
        role="navigation"
      >
        <div
          class="rs-navbar-body"
          style="text-align: center; height: 30px; font-size: 15px;"
        >
          <div
            data-testid="FOOTER_COPYRIGHT_TEXT"
          >
            © Copyright 
            2022
            , SasanLabs
          </div>
        </div>
      </div>
    </footer>
  </div>
</div>
`;
